trie/snap_database.go

```diff
@@ -0,0 +1,509 @@
+// diffs can be applied to rollback. The deepest reorg can be handled depends on
+// the amount of reverse diffs tracked in the disk.
+//
+// At most one readable and writable snap database can be opened at the same time
+// in the whole system which ensures that only one database writer can operate
+// disk state. Unexpected open operations can cause the system to panic.
+type snapDatabase struct {
+	// readOnly is the flag whether the mutation is allowed to be applied.
+	// It will be set automatically when the database is journaled during
+	// the shutdown to reject all following unexpected mutations.
+	readOnly  bool             // Indicator if database is opened in read only mode
+	dirtySize int              // Memory allowance (in bytes) for caching dirty nodes
+	config    *Config          // Configuration for database
+	diskdb    ethdb.Database   // Persistent storage for matured trie nodes
+	cleans    *fastcache.Cache // GC friendly memory cache of clean node RLPs
+	tree      *layerTree       // The group for all known layers
+	freezer   *rawdb.Freezer   // Freezer for storing reverse diffs, nil possible in tests
+	lock      sync.RWMutex     // Lock to prevent mutations from happening at the same time
+}
+
+// openSnapDatabase attempts to load an already existing snapshot from a
+// persistent key-value store (with a number of memory layers from a journal).
+// If the journal is not matched with the base persistent layer, all the
+// recorded diff layers are discarded.
+func openSnapDatabase(diskdb ethdb.Database, cleans *fastcache.Cache, config *Config) *snapDatabase {
+	// Resolve settings from configuration
+	var dirtySize = defaultCacheSize
+	if config != nil && config.DirtySize != 0 {
+		dirtySize = config.DirtySize * 1024 * 1024
+	}
+	readOnly := config != nil && config.ReadOnly
+
+	db := &snapDatabase{
+		readOnly:  readOnly,
+		dirtySize: dirtySize,
+		config:    config,
+		diskdb:    diskdb,
+		cleans:    cleans,
+	}
+	// Construct the layer tree by resolving the in-disk singleton state
+	// and in-memory layer journal.
+	db.tree = newLayerTree(db.loadSnapshot())
+
+	// Open the freezer for reverse diffs if the passed database contains an
+	// ancient store. Otherwise, all the relevant functionalities are disabled.
+	//
+	// Because the freezer can only be opened once at the same time, this
+	// mechanism also ensures that at most one **non-readOnly** snap database
+	// is opened at the same time to prevent accidental mutation.
+	if ancient, err := diskdb.AncientDatadir(); err == nil && ancient != "" && !readOnly {
+		freezer, err := rawdb.NewReverseDiffFreezer(ancient, false)
+		if err != nil {
+			log.Crit("Failed to open reverse diff freezer", "err", err)
+		}
+		db.freezer = freezer
+
+		// Truncate the extra reverse diffs above in freezer in case it's not
+		// aligned with the disk layer.
+		truncateDiffs(db.freezer, diskdb, db.tree.bottom().ID())
```

Comment:
> This can potentially delete a lot of data, right?

More or less YES. In PBSS, whenever a layer is merged into bottom-most disklayer's cache, the reverse diff will be written into freezer, but at that time, the corresponding states are still in memory(disklayer's cache) and can be lost(e.g. panic). At the startup, we need to truncate the extra reverse diffs in case the corresponding states are really lost.

We do have a log for it in case any deletion happens,

```go
log.Info("Truncated extra reverse diff", "number", pruned, "head", target)
```